src/lib.rs

```diff
@@ -0,0 +1,324 @@
+// Copyright 2019. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+mod bindings;
+
+use bindings::{
+    randomx_alloc_cache, randomx_alloc_dataset, randomx_cache, randomx_calculate_hash,
+    randomx_create_vm, randomx_dataset, randomx_dataset_item_count, randomx_destroy_vm,
+    randomx_get_dataset_memory, randomx_init_cache, randomx_init_dataset, randomx_release_cache,
+    randomx_release_dataset, randomx_vm, randomx_vm_set_cache, randomx_vm_set_dataset,
+    RANDOMX_DATASET_ITEM_SIZE, RANDOMX_HASH_SIZE,
+};
+use std::error;
+use std::fmt;
+use std::mem;
+use std::os::raw::{c_char, c_uint, c_ulong, c_void};
+use std::ptr;
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub enum RandomXFlag {
+    FlagDefault,
+    FlagLargePages,
+    FlagHardAES,
+    FlagFullMem,
+    FlagJIT,
+    FlagSecure,
+}
+
+impl RandomXFlag {
+    pub fn value(self) -> c_uint {
+        match self {
+            RandomXFlag::FlagDefault => 0,
+            RandomXFlag::FlagLargePages => 1,
+            RandomXFlag::FlagHardAES => 2,
+            RandomXFlag::FlagFullMem => 4,
+            RandomXFlag::FlagJIT => 8,
+            RandomXFlag::FlagSecure => 16,
+        }
+    }
+}
+
+impl PartialEq for RandomXFlag {
+    fn eq(&self, other: &Self) -> bool {
+        self.value() == other.value()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct CreationError;
```

Comment:
FFR: This is fine, but FYI we've sort of standardized on using the `derive_error` crate for handling all the error boilerplate